<a href="https://colab.research.google.com/github/LS10432/PDDS-PYD-0825-06Oct2025B/blob/main/Project/A2/A2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

# Load the data
df = pd.read_csv('/content/owid-covid-data.csv')
print(f"Number of columns before removal: {df.shape[1]}")
print(f"Total number of rows: {df.shape[0]}")

# Display head and tail
print("\n" + "="*60)
print("Display the first 5 rows of the DataFrame.")
print("\n" + "="*60)
display(df.head())

print("\n" + "="*60)
print("Display the last 5 rows of the DataFrame.")
print("\n" + "="*60)
display(df.tail())

# Check for missing values and identify columns to drop
missing_values_percentage = df.isnull().mean() * 100
columns_to_drop = missing_values_percentage[missing_values_percentage > 90].index.tolist()
print("\nColumns with more than 90% missing values:")
print(columns_to_drop)
print(f"\nNumber of columns to be dropped: {len(columns_to_drop)}")

# Remove columns
df = df.drop(columns=columns_to_drop)

print(f"\nNumber of columns after removal: {df.shape[1]}")
display(df.columns)

# convert 'date' column to datetime data type
df['date'] = pd.to_datetime(df['date'])
print("\n" + "="*60)
print("\nColumn info after datetime conversion:")
df.info()




Number of columns before removal: 67
Total number of rows: 358838

Display the first 5 rows of the DataFrame.



iso_code continent     location        date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2020-01-03          NaN        0.0   
1      AFG      Asia  Afghanistan  2020-01-04          NaN        0.0   
2      AFG      Asia  Afghanistan  2020-01-05          NaN        0.0   
3      AFG      Asia  Afghanistan  2020-01-06          NaN        0.0   
4      AFG      Asia  Afghanistan  2020-01-07          NaN        0.0   

   new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
0                 NaN           NaN         0.0                  NaN  ...   
1                 NaN           NaN         0.0                  NaN  ...   
2                 NaN           NaN         0.0                  NaN  ...   
3                 NaN           NaN         0.0                  NaN  ...   
4                 NaN           NaN         0.0                  NaN  ...   

   male_smokers  handwashing_facilities  hospital_beds_per_thousand  \
0           NaN                  37.746                         0.5   
1           NaN                  37.746                         0.5   
2           NaN                  37.746                         0.5   
3           NaN                  37.746                         0.5   
4           NaN                  37.746                         0.5   

   life_expectancy  human_development_index  population  \
0            64.83                    0.511  41128772.0   
1            64.83                    0.511  41128772.0   
2            64.83                    0.511  41128772.0   
3            64.83                    0.511  41128772.0   
4            64.83                    0.511  41128772.0   

   excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                   NaN                          NaN   
1                                   NaN                          NaN   
2                                   NaN                          NaN   
3                                   NaN                          NaN   
4                                   NaN                          NaN   

   excess_mortality  excess_mortality_cumulative_per_million  
0               NaN                                      NaN  
1               NaN                                      NaN  
2               NaN                                      NaN  
3               NaN                                      NaN  
4               NaN                                      NaN  

[5 rows x 67 columns]


Display the last 5 rows of the DataFrame.



iso_code continent  location        date  total_cases  new_cases  \
358833      ZWE    Africa  Zimbabwe  2023-11-18     265890.0        0.0   
358834      ZWE    Africa  Zimbabwe  2023-11-19     265890.0        0.0   
358835      ZWE    Africa  Zimbabwe  2023-11-20     265890.0        0.0   
358836      ZWE    Africa  Zimbabwe  2023-11-21     265890.0        0.0   
358837      ZWE    Africa  Zimbabwe  2023-11-22     265890.0        0.0   

        new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  \
358833                 0.0        5725.0         0.0                  0.0   
358834                 0.0        5725.0         0.0                  0.0   
358835                 0.0        5725.0         0.0                  0.0   
358836                 0.0        5725.0         0.0                  0.0   
358837                 0.0        5725.0         0.0                  0.0   

        ...  male_smokers  handwashing_facilities  hospital_beds_per_thousand  \
358833  ...          30.7                  36.791                         1.7   
358834  ...          30.7                  36.791                         1.7   
358835  ...          30.7                  36.791                         1.7   
358836  ...          30.7                  36.791                         1.7   
358837  ...          30.7                  36.791                         1.7   

        life_expectancy  human_development_index  population  \
358833            61.49                    0.571  16320539.0   
358834            61.49                    0.571  16320539.0   
358835            61.49                    0.571  16320539.0   
358836            61.49                    0.571  16320539.0   
358837            61.49                    0.571  16320539.0   

        excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
358833                                   NaN                          NaN   
358834                                   NaN                          NaN   
358835                                   NaN                          NaN   
358836                                   NaN                          NaN   
358837                                   NaN                          NaN   

        excess_mortality  excess_mortality_cumulative_per_million  
358833               NaN                                      NaN  
358834               NaN                                      NaN  
358835               NaN                                      NaN  
358836               NaN                                      NaN  
358837               NaN                                      NaN  

[5 rows x 67 columns]


Columns with more than 90% missing values:
['weekly_icu_admissions', 'weekly_icu_admissions_per_million', 'weekly_hosp_admissions', 'weekly_hosp_admissions_per_million', 'excess_mortality_cumulative_absolute', 'excess_mortality_cumulative', 'excess_mortality', 'excess_mortality_cumulative_per_million']

Number of columns to be dropped: 8

Number of columns after removal: 59


Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'new_cases_smoothed', 'total_deaths', 'new_deaths',
       'new_deaths_smoothed', 'total_cases_per_million',
       'new_cases_per_million', 'new_cases_smoothed_per_million',
       'total_deaths_per_million', 'new_deaths_per_million',
       'new_deaths_smoothed_per_million', 'reproduction_rate', 'icu_patients',
       'icu_patients_per_million', 'hosp_patients',
       'hosp_patients_per_million', 'total_tests', 'new_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'positive_rate', 'tests_per_case', 'tests_units', 'total_vaccinations',
       'people_vaccinated', 'people_fully_vaccinated', 'total_boosters',
       'new_vaccinations', 'new_vaccinations_smoothed',
       'total_vaccinations_per_hundred', 'people_vaccinated_per_hundred',
       'people_fully_vaccinated_per_hundred', 'total_boosters_per_hundred',



Column info after datetime conversion:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 358838 entries, 0 to 358837
Data columns (total 59 columns):
 #   Column                                      Non-Null Count   Dtype         
---  ------                                      --------------   -----         
 0   iso_code                                    358838 non-null  object        
 1   continent                                   341730 non-null  object        
 2   location                                    358838 non-null  object        
 3   date                                        358838 non-null  datetime64[ns]
 4   total_cases                                 320628 non-null  float64       
 5   new_cases                                   349032 non-null  float64       
 6   new_cases_smoothed                          347773 non-null  float64       
 7   total_deaths                                299086 non-null  float64       
 8   new_deaths                       

# Activity 2: Data Cleaning and Feature Engineering
Clean the data by removing duplicate rows, imputing missing values, and creating new features. Then, explore the unique countries in the dataset.

## Remove duplicate rows

### Subtask:
Remove duplicate rows from the DataFrame and show the count of rows before and after removal, as well as the number of duplicate and unique duplicate rows.


**Reasoning**:
I need to remove duplicate rows from the DataFrame and show the count of rows before and after removal, as well as the number of duplicate and unique duplicate rows. I will perform all the steps in a single code block.



In [ ]:
# Get the number of rows before removing duplicates
rows_before = df.shape[0]

# Count the number of duplicate rows
duplicate_rows = df.duplicated().sum()

# Count the number of unique duplicate rows
unique_duplicate_rows = df[df.duplicated()].shape[0]

# Remove duplicate rows
df = df.drop_duplicates()

# Get the number of rows after removing duplicates
rows_after = df.shape[0]

# Print the results
print(f"Number of rows before removing duplicates: {rows_before}")
print(f"Number of duplicate rows: {duplicate_rows}")
print(f"Number of unique duplicate rows: {unique_duplicate_rows}")
print(f"Number of rows after removing duplicates: {rows_after}")

Number of rows before removing duplicates: 358838
Number of duplicate rows: 0
Number of unique duplicate rows: 0
Number of rows after removing duplicates: 358838


## Impute missing values

### Subtask:
Impute missing values in the dataset columns using an appropriate method and explain the reasoning.


**Reasoning**:
Impute missing values in numerical and categorical columns using mean and mode respectively, then verify the imputation by checking for remaining missing values.



In [ ]:
# Print the number of missing values per column before imputation
print("Number of missing values per column before imputation:")
print(df.isnull().sum())


# Select numerical columns
numerical_cols = df.select_dtypes(include=['float64', 'int64']).columns

# Impute missing values in numerical columns using the mean
for col in numerical_cols:
    if df[col].isnull().any():
        mean_val = df[col].mean()
        df[col] = df[col].fillna(mean_val)

# Select categorical columns
categorical_cols = df.select_dtypes(include=['object']).columns

# Impute missing values in categorical columns using the mode
for col in categorical_cols:
    if df[col].isnull().any():
        mode_val = df[col].mode()[0] # mode() can return multiple values, take the first
        df[col] = df[col].fillna(mode_val)


# Print the number of remaining missing values per column
print("\nNumber of remaining missing values per column after imputation:")
print(df.isnull().sum())

# Reasoning for imputation methods
print("\nReasoning for imputation methods:")
print("Mean is used for numerical columns because it represents the central tendency of the data and is a common method for imputing missing numerical values when the data is not heavily skewed.")
print("Mode is used for categorical columns because it represents the most frequent category, which is a suitable way to fill missing values in discrete data where the concept of mean is not applicable.")

Number of missing values per column before imputation:
iso_code                                           0
continent                                      17108
location                                           0
date                                               0
total_cases                                    38210
new_cases                                       9806
new_cases_smoothed                             11065
total_deaths                                   59752
new_deaths                                      9749
new_deaths_smoothed                            10979
total_cases_per_million                        38210
new_cases_per_million                           9806
new_cases_smoothed_per_million                 11065
total_deaths_per_million                       59752
new_deaths_per_million                          9749
new_deaths_smoothed_per_million                10979
reproduction_rate                             174021
icu_patients                                

## Create new features

### Subtask:
Extract year and month from the 'date' column, explain the reasoning, and what are the uses.


**Reasoning**:
Extract year and month from the 'date' column and explain the utility of this process.



In [ ]:
# Extract year and month from the 'date' column
df['year'] = df['date'].dt.year
df['month'] = df['date'].dt.month

# Explain the reasoning
print("\nReasoning for extracting year and month:")
print("Extracting year and month from the date column is crucial for time-series analysis. It allows for grouping data by specific time periods, identifying seasonal trends, and comparing metrics across different years and months. This granularity in time data is essential for understanding patterns and changes in the COVID-19 data over time.")

# Display the head of the DataFrame to show the new columns
print("\n" + "="*60)
print("Display the first 5 rows of the DataFrame with 'year' and 'month' columns.")
print("\n" + "="*60)
display(df.head())


Reasoning for extracting year and month:
Extracting year and month from the date column is crucial for time-series analysis. It allows for grouping data by specific time periods, identifying seasonal trends, and comparing metrics across different years and months. This granularity in time data is essential for understanding patterns and changes in the COVID-19 data over time.

Display the first 5 rows of the DataFrame with 'year' and 'month' columns.



iso_code continent     location       date   total_cases  new_cases  \
0      AFG      Asia  Afghanistan 2020-01-03  6.862419e+06        0.0   
1      AFG      Asia  Afghanistan 2020-01-04  6.862419e+06        0.0   
2      AFG      Asia  Afghanistan 2020-01-05  6.862419e+06        0.0   
3      AFG      Asia  Afghanistan 2020-01-06  6.862419e+06        0.0   
4      AFG      Asia  Afghanistan 2020-01-07  6.862419e+06        0.0   

   new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
0         9409.255662  86954.368292         0.0            83.807025  ...   
1         9409.255662  86954.368292         0.0            83.807025  ...   
2         9409.255662  86954.368292         0.0            83.807025  ...   
3         9409.255662  86954.368292         0.0            83.807025  ...   
4         9409.255662  86954.368292         0.0            83.807025  ...   

   diabetes_prevalence  female_smokers  male_smokers  handwashing_facilities  \
0                 9.59       10.793543     32.910129                  37.746   
1                 9.59       10.793543     32.910129                  37.746   
2                 9.59       10.793543     32.910129                  37.746   
3                 9.59       10.793543     32.910129                  37.746   
4                 9.59       10.793543     32.910129                  37.746   

   hospital_beds_per_thousand  life_expectancy  human_development_index  \
0                         0.5            64.83                    0.511   
1                         0.5            64.83                    0.511   
2                         0.5            64.83                    0.511   
3                         0.5            64.83                    0.511   
4                         0.5            64.83                    0.511   

   population  year  month  
0  41128772.0  2020      1  
1  41128772.0  2020      1  
2  41128772.0  2020      1  
3  41128772.0  2020      1  
4  41128772.0  2020      1  

[5 rows x 61 columns]

## Explore unique countries

### Subtask:
Explore the unique countries in the dataset and count the total number of countries.


**Reasoning**:
Get the unique values from the 'location' column and count them to find the number of unique countries.



In [ ]:
# Get unique countries from the 'location' column
unique_countries = df['location'].unique()

# Count the number of unique countries
num_unique_countries = len(unique_countries)

# Print the total count of unique countries
print(f"Total number of unique countries: {num_unique_countries}")

# Display the list of unique countries (optional, for verification)
print("\nList of unique countries:")
print(unique_countries)

Total number of unique countries: 255

List of unique countries:
['Afghanistan' 'Africa' 'Albania' 'Algeria' 'American Samoa' 'Andorra'
 'Angola' 'Anguilla' 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Aruba'
 'Asia' 'Australia' 'Austria' 'Azerbaijan' 'Bahamas' 'Bahrain'
 'Bangladesh' 'Barbados' 'Belarus' 'Belgium' 'Belize' 'Benin' 'Bermuda'
 'Bhutan' 'Bolivia' 'Bonaire Sint Eustatius and Saba'
 'Bosnia and Herzegovina' 'Botswana' 'Brazil' 'British Virgin Islands'
 'Brunei' 'Bulgaria' 'Burkina Faso' 'Burundi' 'Cambodia' 'Cameroon'
 'Canada' 'Cape Verde' 'Cayman Islands' 'Central African Republic' 'Chad'
 'Chile' 'China' 'Colombia' 'Comoros' 'Congo' 'Cook Islands' 'Costa Rica'
 "Cote d'Ivoire" 'Croatia' 'Cuba' 'Curacao' 'Cyprus' 'Czechia'
 'Democratic Republic of Congo' 'Denmark' 'Djibouti' 'Dominica'
 'Dominican Republic' 'Ecuador' 'Egypt' 'El Salvador' 'England'
 'Equatorial Guinea' 'Eritrea' 'Estonia' 'Eswatini' 'Ethiopia' 'Europe'
 'European Union' 'Faeroe Islands' 'Falkland Islands

## Summary:

### Data Analysis Key Findings

*   The initial dataset contained 358,838 rows, and no duplicate rows were found.
*   Missing values in numerical columns were imputed using the mean of each column, and missing values in categorical columns were imputed using the mode of each column. After imputation, there were no remaining missing values.
*   New features 'year' and 'month' were successfully extracted from the 'date' column.
*   There are 255 unique locations in the dataset, including countries, continents, income levels, and the 'World' aggregate.

### Insights or Next Steps

*   The data is clean and ready for further analysis, including time-series analysis using the newly created 'year' and 'month' features.
*   Consider exploring the different types of locations (countries, continents, etc.) within the 'location' column for more granular analysis.


# Consolidated

In [21]:
import pandas as pd

# Load the data
df = pd.read_csv('/content/owid-covid-data.csv')
print(f"Number of columns before removal: {df.shape[1]}")
print(f"Total number of rows: {df.shape[0]}")

# Check for missing values and identify columns to drop
missing_values_percentage = df.isnull().mean() * 100
columns_to_drop = missing_values_percentage[missing_values_percentage > 90].index.tolist()
print("\nColumns with more than 90% missing values:")
print(columns_to_drop)
print(f"\nNumber of columns to be dropped: {len(columns_to_drop)}")

# Remove columns
df = df.drop(columns=columns_to_drop)

print(f"\nNumber of columns after removal: {df.shape[1]}")
display(df.columns)

# convert 'date' column to datetime data type
df['date'] = pd.to_datetime(df['date'])
print("\n" + "="*60)
print("\nColumn info after datetime conversion:")
df.info()

# Remove duplicate rows
rows_before = df.shape[0]
duplicate_rows = df.duplicated().sum()
unique_duplicate_rows = df[df.duplicated()].shape[0]
df = df.drop_duplicates()
rows_after = df.shape[0]

print(f"\nNumber of rows before removing duplicates: {rows_before}")
print(f"Number of duplicate rows: {duplicate_rows}")
print(f"Number of unique duplicate rows: {unique_duplicate_rows}")
print(f"Number of rows after removing duplicates: {rows_after}")


# Print the number of missing values per column before imputation
print("\nNumber of missing values per column before imputation:")
print(df.isnull().sum())

# Select numerical columns
numerical_cols = df.select_dtypes(include=['float64', 'int64']).columns

# Impute missing values in numerical columns using the mean
for col in numerical_cols:
    if df[col].isnull().any():
        mean_val = df[col].mean()
        df[col] = df[col].fillna(mean_val)

# Select categorical columns
categorical_cols = df.select_dtypes(include=['object']).columns

# Impute missing values in categorical columns using the mode
for col in categorical_cols:
    if df[col].isnull().any():
        mode_val = df[col].mode()[0] # mode() can return multiple values, take the first
        df[col] = df[col].fillna(mode_val)


# Print the number of remaining missing values per column
print("\nNumber of remaining missing values per column after imputation:")
print(df.isnull().sum())

# Reasoning for imputation methods
print("\nReasoning for imputation methods:")
print("Mean is used for numerical columns because it represents the central tendency of the data and is a common method for imputing missing numerical values when the data is not heavily skewed.")
print("Mode is used for categorical columns because it represents the most frequent category, which is a suitable way to fill missing values in discrete data where the concept of mean is not applicable.")


# Extract year and month from the 'date' column
df['year'] = df['date'].dt.year
df['month'] = df['date'].dt.month

# Explain the reasoning
print("\nReasoning for extracting year and month:")
print("Extracting year and month from the date column is crucial for time-series analysis. It allows for grouping data by specific time periods, identifying seasonal trends, and comparing metrics across different years and months. This granularity in time data is essential for understanding patterns and changes in the COVID-19 data over time.")

# Get unique countries from the 'location' column
unique_countries = df['location'].unique()

# Count the number of unique countries
num_unique_countries = len(unique_countries)

# Print the total count of unique countries
print(f"\nTotal number of unique countries: {num_unique_countries}")

# Display the list of unique countries (optional, for verification)
print("\nList of unique countries:")
print(unique_countries)

Number of columns before removal: 67
Total number of rows: 358838

Columns with more than 90% missing values:
['weekly_icu_admissions', 'weekly_icu_admissions_per_million', 'weekly_hosp_admissions', 'weekly_hosp_admissions_per_million', 'excess_mortality_cumulative_absolute', 'excess_mortality_cumulative', 'excess_mortality', 'excess_mortality_cumulative_per_million']

Number of columns to be dropped: 8

Number of columns after removal: 59


Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'new_cases_smoothed', 'total_deaths', 'new_deaths',
       'new_deaths_smoothed', 'total_cases_per_million',
       'new_cases_per_million', 'new_cases_smoothed_per_million',
       'total_deaths_per_million', 'new_deaths_per_million',
       'new_deaths_smoothed_per_million', 'reproduction_rate', 'icu_patients',
       'icu_patients_per_million', 'hosp_patients',
       'hosp_patients_per_million', 'total_tests', 'new_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'positive_rate', 'tests_per_case', 'tests_units', 'total_vaccinations',
       'people_vaccinated', 'people_fully_vaccinated', 'total_boosters',
       'new_vaccinations', 'new_vaccinations_smoothed',
       'total_vaccinations_per_hundred', 'people_vaccinated_per_hundred',
       'people_fully_vaccinated_per_hundred', 'total_boosters_per_hundred',



Column info after datetime conversion:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 358838 entries, 0 to 358837
Data columns (total 59 columns):
 #   Column                                      Non-Null Count   Dtype         
---  ------                                      --------------   -----         
 0   iso_code                                    358838 non-null  object        
 1   continent                                   341730 non-null  object        
 2   location                                    358838 non-null  object        
 3   date                                        358838 non-null  datetime64[ns]
 4   total_cases                                 320628 non-null  float64       
 5   new_cases                                   349032 non-null  float64       
 6   new_cases_smoothed                          347773 non-null  float64       
 7   total_deaths                                299086 non-null  float64       
 8   new_deaths                       